tests/abi/tools/revng-check-compatibility-with-abi/Verify.cpp

```diff
@@ -230,114 +227,299 @@ void VH::verify(const abi::FunctionType::Layout &FunctionLayout,
+                              "mentioned in the definition of '"
+                            + std::string(ABI.getName()) + "' abi?";
+        revng_abort(Error.c_str());
+      }
+    }
+  }
+}
+
+static abi::FunctionType::Layout
+getPrototypeLayout(const model::Function &Function, model::ABI::Values ABI) {
+  const model::Type *Prototype = Function.Prototype().getConst();
+  revng_assert(Prototype != nullptr);
+  if (auto *CABI = llvm::dyn_cast<model::CABIFunctionType>(Prototype)) {
+    // Copy the prototype since we might have to modify it before testing.
+    model::CABIFunctionType PrototypeCopy = *CABI;
+
+    if (ABI != PrototypeCopy.ABI()) {
+      // Workaround for dwarf sometimes misdetecting a very specific ABI,
+      // for example, it labels `SystemV_x86_regparam_N` as just `SystemV_x86`
+      // despite them being incompatible.
+      std::array AllowedABIs = { model::ABI::SystemV_x86_regparm_3,
+                                 model::ABI::SystemV_x86_regparm_2,
+                                 model::ABI::SystemV_x86_regparm_1 };
+      revng_assert(llvm::is_contained(AllowedABIs, ABI));
+      PrototypeCopy.ABI() = ABI;
     }
+
+    return abi::FunctionType::Layout(PrototypeCopy);
+  } else if (auto *Raw = llvm::dyn_cast<model::RawFunctionType>(Prototype)) {
+    return abi::FunctionType::Layout(*Raw);
+  } else {
+    revng_abort("Layouts of non-function types are not supported.");
   }
 }
 
 void verifyABI(const TupleTree<model::Binary> &Binary,
                llvm::StringRef RuntimeArtifact,
                model::ABI::Values ABI) {
   model::Architecture::Values Architecture = model::ABI::getArchitecture(ABI);
-  auto ParsedArtifact = abi::artifact::parse(RuntimeArtifact, Architecture);
+  auto Parsed = abi::runtime_test::parse(RuntimeArtifact, Architecture);
 
   llvm::StringRef ArchitectureName = model::Architecture::getName(Architecture);
-  revng_assert(ArchitectureName == ParsedArtifact.Architecture);
+  revng_check(ArchitectureName == Parsed.Architecture);
 
   const abi::Definition &Def = abi::Definition::get(ABI);
-  VerificationHelper Helper{ Architecture, Def, ParsedArtifact.IsLittleEndian };
+  VerificationHelper Helper{ Architecture, Def, Parsed.IsLittleEndian };
+  auto LookupHelper = [&Helper](const auto &Test) {
+    return Test.Name == Helper.FunctionName;
+  };
 
-  std::size_t TestedCount = 0;
+  std::size_t ArgumentTestCount = 0, ReturnValueTestCount = 0;
   for (auto &Function : Binary->Functions()) {
-    const std::string &Name = Function.OriginalName();
-    auto CurrentFunction = ParsedArtifact.Functions.find(Name);
-    if (CurrentFunction == ParsedArtifact.Functions.end()) {
-      // Ignore types not present in the artifact.
+    Helper.FunctionName = Function.OriginalName();
+    if (Helper.FunctionName.take_front(5) == "test_")
+      Helper.FunctionName = Helper.FunctionName.drop_front(5);
+    if (auto Test = llvm::find_if(Parsed.ArgumentTests, LookupHelper);
```

Comment:
`count_if`. Review all usages.

Reply:
Wait, what do you mean? Something like
```diff
+revng_assert(llvm::count_if(Parsed.ArgumentTests, LookupHelper) <= 1);
if (auto Test = llvm::find_if(Parsed.ArgumentTests, LookupHelper);
    Test != Parsed.ArgumentTests.end()) {
```
?

Reply:
`find_if(...) != end` is the same as `count_if(...) != 0`. Prefer the second syntax.

Reply:
The `Test` variable is used inside the body of the loop. It's doing the search for a reason, not to just check if it exists.

P. S. consider that after dropping `SortedVector` this becomes `std::map::find`.